src/Razor/test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Cohost/CohostCodeActionsEndpointTest.cs

```diff
@@ -870,6 +872,150 @@ Hello World
+            codeActionName: LanguageServerConstants.CodeActions.PromoteUsingDirective,
+            additionalExpectedFiles: [
+                (FileUri(@"..\_Imports.razor"), """
+                    @using System
+                    """)]);
+    }
+
+    [Fact]
+    public async Task PromoteUsingDirective_Mvc()
+    {
+        await VerifyCodeActionAsync(
+            input: """
+                @using [||]System
+
+                <div>
+                    Hello World
+                </div>
+                """,
+            expected: """
+
+                <div>
+                    Hello World
+                </div>
+                """,
+            codeActionName: LanguageServerConstants.CodeActions.PromoteUsingDirective,
+            fileKind: FileKinds.Legacy,
+            additionalExpectedFiles: [
+                (FileUri(@"..\_ViewImports.cshtml"), """
+                    @using System
+                    """)]);
+    }
+
+    [Fact]
+    public async Task PromoteUsingDirective_ExistingImports()
+    {
+        await VerifyCodeActionAsync(
+            input: """
+                @using [||]System
+
+                <div>
+                    Hello World
+                </div>
+                """,
+            additionalFiles: [
+                (FilePath(@"..\_Imports.razor"), """
+                    @using System.Text
+                    @using Foo.Bar
+                    """)],
+            expected: """
+
+                <div>
+                    Hello World
+                </div>
+                """,
+            codeActionName: LanguageServerConstants.CodeActions.PromoteUsingDirective,
+            additionalExpectedFiles: [
+                (FileUri(@"..\_Imports.razor"), """
+                    @using System.Text
+                    @using Foo.Bar
+                    @using System
+                    """)]);
+    }
+
+    [Fact]
+    public async Task PromoteUsingDirective_ExistingImports_BlankLineAtEnd()
+    {
+        await VerifyCodeActionAsync(
+            input: """
+                @using [||]System
+
+                <div>
+                    Hello World
+                </div>
+                """,
+            additionalFiles: [
+                (FilePath(@"..\_Imports.razor"), """
+                    @using System.Text
+                    @using Foo.Bar
+                    
+                    """)],
+            expected: """
+
+                <div>
+                    Hello World
+                </div>
+                """,
+            codeActionName: LanguageServerConstants.CodeActions.PromoteUsingDirective,
+            additionalExpectedFiles: [
+                (FileUri(@"..\_Imports.razor"), """
+                    @using System.Text
+                    @using Foo.Bar
+                    @using System
+                    """)]);
+    }
+
+    [Fact]
+    public async Task PromoteUsingDirective_ExistingImports_WhitespaceLineAtEnd()
+    {
+        await VerifyCodeActionAsync(
```

Comment:
Is it worth adding a test that showcases moving from Imports/_ViewImports to one a folder up?

Reply:
This is really annoying to do with our current test infra :(

I'll have a go, but will time box it. I wouldn't mind cleaning up some of the code actions infra a bit more, as we add more, so can always come back to it.

Reply:
I'm giving up 🎉

Should be easier once the source generator is hooked up.